Add optional comment to order during checkout

lampyjon opened the pull request:
I want to merge this change because...

This adds the ability for a user to add an optional comment to the order during the checkout process.

Two areas probably require work before merging: is this the best point in the checkout process to capture the note, and would it be useful if the user could see their comment on the completed order summary? 

### Pull Request Checklist

(Please keep this section. It will make maintainer's life easier.)

1. [x] Privileged views and APIs are guarded by proper permission checks.
1. [x] All visible strings are translated with proper context.
1. [x] All data-formatting is locale-aware (dates, numbers, and so on).
1. [x] Database queries are optimized and the number of queries is constant.
1. [x] The changes are tested.
1. [x] The code is documented (docstrings, project documentation).
1. [x] Python code quality checks pass: `pycodestyle`, `pydocstyle`, `pylint`.
1. [x] JavaScript code quality checks pass: `eslint`.

lampyjon commented:
Just realised that I had misunderstood the checkout flow slightly and that this patch doesn't work for carts which don't require shipping.

elwoodxblues commented:
Thanks for the PR @lampyjon. It generally looks very good and such feature could be useful in Saleor, but I'm not convinced that introducing the fourth checkout step is a good idea. Checkout should always be as simple as possible. I think we could show this field in the "summary" step, between "BILLING ADDRESS" and the "Order & pay" button. What do you think?

elwoodxblues commented on saleor/order/models.py:
```diff
@@ -164,6 +164,10 @@ def create_history_entry(self, comment='', status=None, user=None):
             status = self.status
         self.history.create(status=status, comment=comment, user=user)
 
+    def create_order_note(self, note='', user=None):
```
I think this logic is simple enough to be moved directly to the `create_order` function. We try to keep the models simple and avoid additional methods.

lampyjon commented:
Thank you for the feedback @elwoodxblues - I have made changes to the PR as suggested, and added the note field to the final page. I'm not sure if the way I've incorporated the NoteForm into the three views in summary.py is the best way, so would appreciate any feedback on this.

elwoodxblues commented on saleor/order/models.py:
```diff
@@ -383,7 +383,9 @@ def __str__(self):
     user = models.ForeignKey(
-        settings.AUTH_USER_MODEL, on_delete=models.CASCADE)
+        settings.AUTH_USER_MODEL, blank=True, null=True,
+        verbose_name=pgettext_lazy('Order note field', 'user'),
```
We have an issue (#1465) for removing verbose names from models in favor of explicit labels in forms, so we could also remove this `verbose_name`.

elwoodxblues commented:
I've tested it locally and: 
- it doesn't work in the anonymous checkout - `Cannot assign "<SimpleLazyObject: <django.contrib.auth.models.AnonymousUser object at 0x10fe61e10>>": "OrderNote.user" must be a "User" instance.` I think we should allow empty user in OrderNote for this case.
- the migration is failing due to reference to a nonexistent migration "0029_auto_20171228_1752"

Also we should either add new tests or extend the existing checkout tests to check that adding the note in the POST data results in creating a new OrderNote as expected.

lampyjon commented:
Hi @elwoodxblues - I _think_ I've fixed the migration problem; my local system got into a mess when I was jumping around between branches, and so I ended up with the wrong dependency tree / database structure. The anonymous user case is fixed.

elwoodxblues commented:
@lampyjon This looks very good now! Could you fix those Codeclimate issues with trailing spaces and add tests that'll check that OrderNote is properly created? Then this PR will be ready to merge.

lampyjon commented:
Hi @elwoodxblues - CodeClimate warnings should be fixed (other than the complexity warning on summary_with_shipping_view()). 

I've spent an evening trying to figure out pytest, how it works, and what sort of test you are looking for for OrderNote, and I have to admit I'm baffled by it as a framework - is there a good resource you can recommend on this, as I feel I'm going around in circles on it? Thanks!

elwoodxblues commented:
Well, the best would be to take a look at our tests e.g. in `test_checkout.py` and `test_checkout_integration.py` files. Tests that could be added:
- test that assigns some value to `checkout.note` and calls `checkout.create_order()` and checks that OrderNote was created as expected
- tests for three views that you/ve modified that check if including `note` field in the POST data results in setting the `checkout.note` field.

If you have time and want to give it a try, you can try to add those tests. If not, we can try to do it later or in a separate PR.

elwoodxblues commented on tests/test_checkout.py:
```diff
@@ -218,3 +224,37 @@ def test_checkout_create_order_insufficient_stock(
+    request = rf.post('/test/', {'note': 'test_note'})
+    request.user = Mock()
+    request.country = Mock()
+    mock_BillingWithoutShippingAddressForm().is_valid.return_value = False
```
Can we use `lower_case_with_underscores` naming style here?

patrys replied:
I'd also consider:

* always passing a `spec` to `Mock` instances, otherwise they have all possible parameters and we may be masking a bug in the tested code
* using the `monkeypatch` fixture instead of `mock.patch` in cases where it results in less code

elwoodxblues commented:
Thanks @lampyjon and @akjanik for working on this!
